components/google_sheets/actions/upsert-row/upsert-row.mjs

```diff
@@ -0,0 +1,295 @@
+      optional: true,
+    },
+    updates: {
+      type: "object",
+      label: "Update",
+      description: "Update statment: if the spreadsheet contains duplicate key **Value** in some row in specified **Column**, individual cells in the *first* duplicate row will be updated using this object's column-value pairs. Enter the column name for the key and the corresponding column value. You may also disable structured mode to pass a JSON object with key/value pairs representing columns and values (e.g. `{{ { A: 5, B: \"test\" } }}`).",
+      optional: true,
+    },
+  },
+  methods: {
+    /**
+     * Gets the ID of a worksheet in a spreadsheet
+     * @param {string} sheetName - The name of the worksheet
+     * @param {string} spreadsheetId - ID of the spreadsheet
+     * @returns {string} The ID of the worksheet
+     */
+    async getSheetId(sheetName = this.sheetName, spreadsheetId = this.sheetId) {
+      const fields = [
+        "sheets.properties.sheetId",
+      ];
+      return (await this.googleSheets.getSpreadsheet(spreadsheetId, fields, {
+        ranges: [
+          sheetName,
+        ],
+        includeGridData: false,
+      })).sheets[0].properties.sheetId;
+    },
+    /**
+     * https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets/request#UpdateCellsRequest
+     *
+     * @param {string} sheetId - ID of the worksheet
+     * @param {number} row - The row number (>=1) of the cell to update
+     * @param {string} column - The column letter of the cell to update
+     * @param {string} value - The new value of the cell
+     * @returns An UpdateCellsRequest object
+     */
+    getUpdateRequestObj(sheetId, row, column, value) {
+      const colIndex = this.googleSheets._getColumnIndex(column) - 1;
+      return {
+        "updateCells":
+        {
+          "range": {
+            "sheetId": sheetId,
+            "startRowIndex": row - 1,
+            "endRowIndex": row,
+            "startColumnIndex": colIndex,
+            "endColumnIndex": colIndex + 1,
+          },
+          "rows": [
+            {
+              "values": [
+                {
+                  "userEnteredValue": {
+                    "stringValue": value,
+                  },
+                },
+              ],
+            },
+          ],
+          "fields": "*",
+        },
+      };
+    },
+    /**
+     * https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets/request#AddSheetRequest
+     *
+     * Creates a worksheet and returns the properties of the newly created sheet
+     * @param {string} spreadsheetId - ID of the spreadsheet in which to create a worksheet
+     * @param {object} properties - The properties the new sheet should have
+     * @returns An object containing the SheetProperties (`properties`) of the newly created sheet
+     */
+    async addSheet(spreadsheetId, properties) {
```

Comment:
Move `addSheet`, `deleteSheet`, `updateRowCells`, `updateRow`, and `getUpdateRequestObj` methods to the app file for [reusability](https://pipedream.com/docs/components/guidelines/#methods)?